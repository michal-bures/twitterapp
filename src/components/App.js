import React from 'react';
import { Modal, Panel, Form, Glyphicon } from 'react-bootstrap';

import Immutable from 'immutable';
import SearchBox from './SearchBox';
import { connect } from 'react-redux';
import FilterEditor from './FilterEditor';
import Stats from './Stats';
import SortableTable from './SortableTable';
import { showStats, hideStats, setFilter } from '../actions';

// Root application component
let App = ({
    dispatch,
    statsModalVisible,
    filters,
    filterValues,
    tweets,
    onShowStats,
    onCloseStats,
    onSetFilter,
}) => {

    const renderResults=() => {
        if (!tweets) {
            // currently loading data - show nothing
            return;
        } else {
            // display results
            // apply filters to the tweet list
            var displayList = tweets.filter((item) => {
                // ok only if all filters match
                return filters.every((filter, i) => {
                    if (!filter.apply(item, filterValues.get(i))) {
                        return false;
                    }
                    return true;
                });
            })
        

            return (<div>
                <Panel header="Filters" collapsible>
                    <Form horizontal bsSize="small">
                    {filters.map((filter, key) => { return (
                        <FilterEditor key={key}
                            label={filter.label}
                            value={filterValues.get(key)}
                            placeholder={filter.prompt}
                            onChange={(value) => {
                                onSetFilter(key, value);
                            }}
                            validator={filter.validateValue.bind(filter)}
                        />
                    )})}
                    </Form>
                </Panel>
                <a href="#" onClick={onShowStats}>Show statistics</a>

                <SortableTable 
                    columns={Immutable.List.of(
                        {id:"id", label:"#"},
                        {id:"text", label:"text"},
                        {id:"favs", label:<Glyphicon glyph="star"/>},
                        {id:"date", label:"sent", display:"dateString"}
                    )}
                    records={displayList}/>
                <Modal show={statsModalVisible} onHide={onCloseStats}>
                    <Modal.Header closeButton>
                        <Modal.Title>Statistics</Modal.Title>
                    </Modal.Header>
                    <Modal.Body>
                        <Stats tweets={tweets}/>
                    </Modal.Body>
                </Modal>
                </div>
                
            );            
        }

    }

    return (
        <div className="container">
            <h2>Twitter browser demo</h2>
            <SearchBox/>
            {renderResults()}
        </div>
    );
}

App.propTypes = {
    // sequence of available Filters (see filters)
    filters: React.PropTypes.instanceOf(Immutable.List),

    //
    filterValues: React.PropTypes.instanceOf(Immutable.Map),
}

const mapStateToProps = (state) => {
  return {
    tweets: state.get('tweets'),
    statsModalVisible : state.get('statsModalVisible'),
    filters: state.get('filters'),
    filterValues : state.get('filterValues'),
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    onShowStats: () => {
        dispatch(showStats());
    },
    onCloseStats: () => {
        dispatch(hideStats());
    },
    onSetFilter: (filterId, value) => {
        dispatch(setFilter(filterId, value));
    }
  }
}

App = connect(
    mapStateToProps,
    mapDispatchToProps
)(App);

export default App;
