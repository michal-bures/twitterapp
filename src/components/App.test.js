import React from 'react';
import App from './App';
import moment from 'moment';
import { mount } from 'enzyme';
import Immutable from 'immutable';
import twitterApp from '../reducers';
import { Provider } from 'react-redux'
import { createStore } from 'redux'

function mockTweet(id, text, favs, time) {
    let mentions = text.match(/@[^@ ]+/g) || [];
    let hashtags = text.match(/#[^# ]+/g) || [];
    return {
        id: id,
        text: text,
        favs: favs,
        date: moment('2016-01-01T'+time),
        dateString: '1.1.2016 '+time,
        mentions: mentions.map(str => str.slice(1)),
        hashtags: hashtags.map(str => str.slice(1)),
    }
};

describe('<App />', () => {
    
    let i = 0;

    const mockTweets = {
        first : mockTweet(i++,'It\'s never too early for some twitter drama!',0,'3:00'),
        oneMention : mockTweet(i++,'Tweet with one @mention',0,'9:00'),
        oneHashtag : mockTweet(i++,'I can use hastags #awesome',1,'9:15'),
        mostPopular : mockTweet(i++,'Like this if you approve',112,'9:30'),
        threeMentions : mockTweet(i++,'@thatOneGuy @thatOtherGuy @thatOneToo I see!',0,'10:00'),
        threeHashtags : mockTweet(i++,'The edgiest of tweets #sobrave #verytrendy #needsmorehashtags ',0,'10:00'),
    }
    const mockTweetsSequence = Immutable.List.of(
        mockTweets.first, mockTweets.oneMention, mockTweets.oneHashtag, mockTweets.mostPopular,
        mockTweets.threeMentions, mockTweets.threeHashtags
    )

    let wApp;
    let wTable;
    
    beforeEach(() => {
        let store = createStore(twitterApp,
        Immutable.fromJS({
            fetching : false,
            tweets : mockTweetsSequence,
            statsModalVisible : false,
            error: null,
            // sequence filter types
            filters: [],
            // active filter values (key is index of filter in filters)
            filterValues : Immutable.Map()})
        ,);

        wApp = mount(<Provider store={store}><App /></Provider>);

        //console.log(Object.values(mockTweets));
        wTable = wApp.find('SortableTable');
    })

    // DOM navigation helpers
    const tableRowsCount = () => wTable.find('tr').length - 1; // substract one for headers row
    const firstRowText = () => wTable.find('.sortable-table-row-0').find('td').at(1).text();
    const columnHeader = i => { wTable.find('th').at(i) }

    // tests

    it('renders as expected', () => {
        expect(wApp.debug()).toMatchSnapshot();
    });

    it('displays all tweets by default', () => {
        // table should have 5 rows
        expect(tableRowsCount()).toBe(6);
        // test that first table row has expected content
        expect(firstRowText()).toBe(mockTweets.first.text);
    });

    it('can be sorted by clicking on table header', () => {
        // two clicks on "likes" column => now should be sorted by likes in descending order
        columnHeader(2).simulate('click');
        columnHeader(2).simulate('click');
        expect(firstRowText()).toBe(mockTweets.mostPopular.text);
    });

});

