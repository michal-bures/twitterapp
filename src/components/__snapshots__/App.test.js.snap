exports[`<App /> renders as expected 1`] = `
"<Provider store={{...}}>
  <Connect(App)>
    <App tweets={{...}} statsModalVisible={false} filters={{...}} filterValues={{...}} onShowStats={[Function]} onCloseStats={[Function]} onSetFilter={[Function]}>
      <div className=\"container\">
        <h2>
          Twitter browser demo
        </h2>
        <Connect(SearchBox)>
          <SearchBox placeholder=\"Start by entering username\" fetching={[undefined]} error={[undefined]} onSubmit={[Function]} onChange={[Function]}>
            <form onSubmit={[Function]}>
              <FormGroup bsClass=\"form-group\">
                <div className=\"form-group\">
                  <InputGroup bsClass=\"input-group\">
                    <span className=\"input-group\">
                      <FormControl type=\"text\" placeholder=\"Start by entering username\" onChange={[Function]} componentClass=\"input\" bsClass=\"form-control\">
                        <input placeholder=\"Start by entering username\" onChange={[Function]} type=\"text\" id={[undefined]} className=\"form-control\" />
                      </FormControl>
                      <InputGroupButton bsClass=\"input-group-btn\">
                        <span className=\"input-group-btn\">
                          <Button type=\"submit\" bsStyle=\"primary\" disabled={false} active={false} block={false} bsClass=\"btn\">
                            <button type=\"submit\" disabled={false} className=\"btn btn-primary\">
                              <Glyphicon glyph=\"search\" bsClass=\"glyphicon\">
                                <span className=\"glyphicon glyphicon-search\" />
                              </Glyphicon>
                            </button>
                          </Button>
                        </span>
                      </InputGroupButton>
                    </span>
                  </InputGroup>
                </div>
              </FormGroup>
            </form>
          </SearchBox>
        </Connect(SearchBox)>
        <div>
          <Panel header=\"Filters\" collapsible={true} defaultExpanded={false} bsStyle=\"default\" bsClass=\"panel\">
            <div className=\"panel panel-default\" id={{...}}>
              <div className=\"panel-heading\">
                <h4 role=\"presentation\" className=\"panel-title\">
                  <a role={[undefined]} href={[undefined]} onClick={[Function]} aria-controls={[undefined]} aria-expanded={false} aria-selected={false}>
                    Filters
                  </a>
                </h4>
              </div>
              <Collapse in={false} onEnter={[undefined]} onEntering={[undefined]} onEntered={[undefined]} onExit={[undefined]} onExiting={[undefined]} onExited={[undefined]} timeout={300} unmountOnExit={false} transitionAppear={false} dimension=\"height\" getDimensionValue={[Function]}>
                <Transition in={false} onExited={[Function]} timeout={300} unmountOnExit={false} transitionAppear={false} aria-expanded={{...}} className=\"\" exitedClassName=\"collapse\" exitingClassName=\"collapsing\" enteredClassName=\"collapse in\" enteringClassName=\"collapsing\" onEnter={[Function]} onEntering={[Function]} onEntered={[Function]} onExit={[Function]} onExiting={[Function]}>
                  <div id={[undefined]} role={[undefined]} className=\"panel-collapse collapse\" aria-hidden={true} aria-expanded={{...}}>
                    <div className=\"panel-body\">
                      <Form horizontal={true} bsSize=\"small\" inline={false} componentClass=\"form\" bsClass=\"form\">
                        <form className=\"form-horizontal\" />
                      </Form>
                    </div>
                  </div>
                </Transition>
              </Collapse>
            </div>
          </Panel>
          <a href=\"#\" onClick={[Function]}>
            Show statistics
          </a>
          <SortableTable columns={{...}} records={{...}}>
            <Table striped={true} bordered={true} condensed={true} hover={false} responsive={false} bsClass=\"table\">
              <table className=\"table table-striped table-bordered table-condensed\">
                <thead>
                  <tr>
                    <th className=\"clickable no-wrap\" onClick={[Function]}>
                      <Glyphicon glyph=\"chevron-up\" bsClass=\"glyphicon\">
                        <span className=\"glyphicon glyphicon-chevron-up\" />
                      </Glyphicon>
                      #
                    </th>
                    <th className=\"clickable no-wrap\" onClick={[Function]}>
                      text
                    </th>
                    <th className=\"clickable no-wrap\" onClick={[Function]}>
                      <Glyphicon glyph=\"star\" bsClass=\"glyphicon\">
                        <span className=\"glyphicon glyphicon-star\" />
                      </Glyphicon>
                    </th>
                    <th className=\"clickable no-wrap\" onClick={[Function]}>
                      sent
                    </th>
                  </tr>
                </thead>
                <tbody>
                  <tr className=\"sortable-table-row-0\">
                    <td>
                      0
                    </td>
                    <td>
                      It&#39;s never too early for some twitter drama!
                    </td>
                    <td>
                      0
                    </td>
                    <td>
                      1.1.2016 3:00
                    </td>
                  </tr>
                  <tr className=\"sortable-table-row-1\">
                    <td>
                      1
                    </td>
                    <td>
                      Tweet with one @mention
                    </td>
                    <td>
                      0
                    </td>
                    <td>
                      1.1.2016 9:00
                    </td>
                  </tr>
                  <tr className=\"sortable-table-row-2\">
                    <td>
                      2
                    </td>
                    <td>
                      I can use hastags #awesome
                    </td>
                    <td>
                      1
                    </td>
                    <td>
                      1.1.2016 9:15
                    </td>
                  </tr>
                  <tr className=\"sortable-table-row-3\">
                    <td>
                      3
                    </td>
                    <td>
                      Like this if you approve
                    </td>
                    <td>
                      112
                    </td>
                    <td>
                      1.1.2016 9:30
                    </td>
                  </tr>
                  <tr className=\"sortable-table-row-4\">
                    <td>
                      4
                    </td>
                    <td>
                      @thatOneGuy @thatOtherGuy @thatOneToo I see!
                    </td>
                    <td>
                      0
                    </td>
                    <td>
                      1.1.2016 10:00
                    </td>
                  </tr>
                  <tr className=\"sortable-table-row-5\">
                    <td>
                      5
                    </td>
                    <td>
                      The edgiest of tweets #sobrave #verytrendy #needsmorehashtags 
                    </td>
                    <td>
                      0
                    </td>
                    <td>
                      1.1.2016 10:00
                    </td>
                  </tr>
                </tbody>
              </table>
            </Table>
          </SortableTable>
          <Modal show={false} onHide={[Function]} backdrop={true} keyboard={true} autoFocus={true} enforceFocus={true} manager={{...}} renderBackdrop={[Function]} animation={true} dialogComponentClass={[Function]} bsClass=\"modal\">
            <Modal onHide={[Function]} keyboard={true} autoFocus={true} enforceFocus={true} manager={{...}} renderBackdrop={[Function]} show={false} onEntering={[Function]} onExited={[Function]} backdrop={true} backdropClassName=\"modal-backdrop\" containerClassName=\"modal-open\" transition={[Function]} dialogTransitionTimeout={300} backdropTransitionTimeout={150} />
          </Modal>
        </div>
      </div>
    </App>
  </Connect(App)>
</Provider>"
`;
